language/vm/src/abi.rs

```diff
@@ -0,0 +1,399 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access::ModuleAccess,
+    file_format::{
+        CompiledModule, FieldDefinition, FunctionHandle, Kind, SignatureToken, StructDefinition,
+        StructFieldInformation, TypeParameterIndex,
+    },
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::{StructTag, TypeTag},
+};
+
+/// Defines the Application Binary Interface (ABI) for Move modules, the `check_compatibility`
+/// function for checking ABI compatibility of two modules, and the `can_update` function for
+/// determining whether a new version of a module can safely update the old one.
+
+/// A normalized version of `SignatureToken`, a type expression appearing in struct or function
+/// declarations. Unlike `SignatureToken`s, `TypeABI`s from different modules can safely be compared.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TypeABI {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Signer,
+    Struct {
+        address: AccountAddress,
+        module: Identifier,
+        name: Identifier,
+        type_arguments: Vec<TypeABI>,
+    },
+    Vector(Box<TypeABI>),
+    TypeParameter(TypeParameterIndex),
+    Reference(Box<TypeABI>),
+    MutableReference(Box<TypeABI>),
+}
+
+/// ABI for a `FieldDefinition`. The `name` is included even though it is
+/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `StructABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldABI {
+    pub name: Identifier,
+    pub type_: TypeABI,
+}
+
+/// ABI for a kind constraint on a generic type parameter
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum KindABI {
+    All,
+    Resource,
+    Copyable,
+}
```

Comment:
Agreed that this looks a bit silly because this is identical to `file_format::Kind`. I did this because I wanted the normalized types to be completely independent from the types defined in file_format.rs. The discipline I'm going for is that these normalized types are at a higher level of abstraction than the file format and can tolerate non-semantic changes in the more low-level bytecode file format.